### ISSUE TITLE
`typings.py` file inside a project causes pyright to report that this is not a valid directory.

### ISSUE THREAD

**Issue by pawelrubin** [microsoft/pyright#777]
**Describe the bug**
If you create a file inside a project named `typings.py`, pyright in CLI mode will produce an error:
```shell
stubPath <project_directory>/typings is not a valid directory.
```

**To Reproduce**
Steps to reproduce the behavior.

Simply create a `typings.py` file inside your project.

```shell
source/
  main.py
  typings.py
```

However, if a `typings` directory is present, the error does not occur.

**Expected behavior**
There should be no errors.

**Additional context**
- VS Code extension: - 1.1.47
  - settings:
    ```json
		{
  			"pyright.typeCheckingMode": "strict",
  			"pyright.useLibraryCodeForTypes": true
		}
	```
- Python: 3.8.3

**Comment by erictraut** [comment]
This log statement has nothing to do with the absence or presence of a "typings.py" file.

If you don't specify a stubPath (i.e. a path where local type stubs can be found), Pyright defaults to the directory called "typings". (Arguably, it should be called "stubs", but for historical reasons we decided to leave it as "typings".) If you don't have any directory called "typings" in your workspace, Pyright will emit this log statement. Note that it's not an error (i.e. it won't cause the CLI to return a non-zero return code). It's just meant to be informational.

If it bothers you, you can add the following line to your pyrightconfig.json file:
```
"stubPath": ""
```

**Comment by ssbarnea** [comment]
@erictraut In fact the situation become bit worse. If you configure `"stubPath" = ""` inside pyproject.toml, you will get a warning logged:

> [Warn] The stubPath has been specified in both the config file and the client settings. The value in the config file (/Users/ssbarnea/c/a/ansible-compat) will take precedence

I mention that there are no user configured custom stub values in vscode when you get this value, it is just a side effect that vscode has an implicit "typings" value in its settings.

I tried to put the empty string inside vscode settings but in that case it only gets another warning

> [Warn  - 10:33:46] (57604) stubPath typings is not a valid directory.

Basically it seems impossible to configure it to avoid any distracting logging messages.

```
[Error - 10:30:29] (57604) Pyproject file "/Users/ssbarnea/c/a/ansible-lint/pyproject.toml" is missing "[tool.pyright]" section.
[Warn  - 10:30:29] (57604) stubPath /Users/ssbarnea/c/a/ansible-lint/typings is not a valid directory.
[Warn  - 10:30:29] (57604) stubPath /Users/ssbarnea/c/a/ansible-compat/typings is not a valid directory.
[Warn  - 10:30:29] (57604) stubPath typings is not a valid directory.
[Warn  - 10:30:32] (57604) Workspace indexing has hit its upper limit: 2000 files
[Warn  - 10:30:59] (57604) stubPath /Users/ssbarnea/c/a/ansible-lint/typings is not a valid directory.
[Warn  - 10:32:31] (57604) The stubPath has been specified in both the config file and the client settings. The value in the config file (/Users/ssbarnea/c/a/ansible-lint) will take precedence
[Warn  - 10:32:52] (57604) The stubPath has been specified in both the config file and the client settings. The value in the config file (/Users/ssbarnea/c/a/ansible-lint) will take precedence
[Warn  - 10:32:55] (57604) The stubPath has been specified in both the config file and the client settings. The value in the config file (/Users/ssbarnea/c/a/ansible-lint) will take precedence
[Warn  - 10:33:02] (57604) The stubPath has been specified in both the config file and the client settings. The value in the config file (/Users/ssbarnea/c/a/ansible-compat) will take precedence
[Warn  - 10:33:46] (57604) stubPath typings is not a valid directory.
```

**Comment by erictraut** [comment]
This is as designed. This is just a log warning to help diagnose potential problems. You can ignore it if you are not making use of a stub path.

**Comment by ssbarnea** [comment]
How many people are using stub paths? Probably less than 1% or even 0.1%, we could easily check how many repositories have such a folder in them. I am sure that the telemetry could report a real occurence level.

One could argue that the message is misleading and impossible to silence because there is no way to configure vscode settings to avoid displaying some kind of warning about these when you do not make use of them. That is why I consider is to be a bug. The only case where I would see it ok, is if would a debug level message. It would not be ok to be even notice level. 

Just to be clear, "is not a valid directory." is the root issue as there is no way to just avoid it when you don't need/have a stub directory. Even creating an empty folder does not silence it. How about considering empty string a method for disabling this functionality? aka "" being no-stubPath.

**Comment by erictraut** [comment]
I think it's fine as is. The logged text is meant to help users (and us maintainers) diagnose problems when they occur. If you are not experiencing a problem, you can ignore the logs.

If this warning really bothers you, you can set "python.analysis.logLevel" to "Error". This will eliminate all warnings from your log output.

**Comment by stevesimmons** [comment]
Eric, I find your reasoning plays down the confusion it causes to your ordinary users. Of course you understand intimately how pyright and pylance operate and know to ignore the warning. Whereas ordinary users (such as me) start from a very different place.

I found the Python Language Server Output tab, and quite justifiably interpreted "Warn - stubPath typings is not a valid directory." to mean something is wrong with my pylance installation or configuration. I then wasted half an hour trying to figure out what it is, eventually ending up at this GH issue saying it is expected behaviour and "fine as it is".

If pylance is going to routinely emit "warn" log lines that show up in red, the least it can do is add a postscript to the message saying "This is normal, not usually indicating a problem" or similar.

**Comment by erictraut** [comment]
@heejaechang, is there a way for pyright to determine whether the `python.analysis.stubPath` setting has a user-defined value rather than the default value? Apparently, if no user-defined value is present in the settings, VS Code fills in the default and sends pyright a value of `"typings"`. The logging code therefore thinks that the user has explicitly specified a `"typings"` directory and emits a warning if the directory doesn't exist. It pyright were instead to receive an empty value from VS Code, it would silently default to `"typings"` without emitting any warning. That's the preferred behavior. Unfortunately, I don't see a way to do this other than by removing the default value for the `python.analysis.stubPath` setting in the package.json manifest. That doesn't sound like the right answer though. Thoughts?

**Comment by debonte** [comment]
> is there a way for pyright to determine whether the python.analysis.stubPath setting has a user-defined value rather than the default value?

Yes, in the client code using vscode APIs. Specifically, [`WorkspaceConfiguration.inspect`](https://vscode-api.js.org/interfaces/vscode.WorkspaceConfiguration.html#inspect). But I'm not sure how you would communicate that to the server. Maybe add a fake config setting to the settings returned by your `workspace/configuration` request handler. Something like `typingsHasUserValue: true`?

**Comment by erictraut** [comment]
That wouldn't work for pylance though. Pylance doesn't (currently) have a custom client.

**Comment by rchiodo** [comment]
> That wouldn't work for pylance though. Pylance doesn't (currently) have a custom client.

Well actually we do. @heejaechang added this a while back. We control the client side now too.

**Comment by erictraut** [comment]
Ah, I didn't realize that was complete. Is this solution one that you're amenable to? If so, we'd need to add it to both the pylance and pyright client. It seems kind of hacky to me. Any other ideas?

**Comment by rchiodo** [comment]
We could just skip that warning for 'typings'? Or maybe make it a warning only when 'verbose' is specified?

**Comment by debonte** [comment]
Oh, if we see via `inspect` that `stubPath` is set to the default, we could send empty string to the server rather than the default value?

**Comment by erictraut** [comment]
Yeah, I like that idea.

**Comment by erictraut** [comment]
If that solution works for the pylance team, perhaps you could open an issue in the pylance-release issue tracker for this work? When you implement it in the pylance client, you could then copy the logic into the pyright client.

**Comment by debonte** [comment]
Opened https://github.com/microsoft/pylance-release/issues/4917
